### PR TITLE
feat(worker): Push Notifications

### PR DESCRIPTION
**Pull request by alxhub**
Rough draft PR

TODO
- [x] from jeffbcross: add a check inside e2e gulp task to make sure `ngsw-config.json` is present, with basic instructions to get it set up.
- [x] fix typings for `jshashes` module, `ServiceWorkerRegistration`, and `CacheStorage`

### CODE REVIEW

**Comment by jeffbcross** [comment]
@alxhub there are some errors when I run `gulp build` because of missing typings for `CacheStorage` and `ServiceWorkerRegistration`. It causes build failures because TypeScript isn't emitting the `context.ts`. I temporarily worked around it by setting types inside `context.ts`. Do you have some uncommitted typings for those?

Also no `jshashes` typings.

```
[13:45:29] Using gulpfile ~/Projects/mobile-toolkit/service-worker/worker/gulpfile.js
[13:45:29] Starting 'build'...
[13:45:29] Starting 'clean'...
[13:45:29] Finished 'clean' after 182 ms
[13:45:29] Starting 'task:generator:build'...
[13:45:29] Starting 'task:generator:compile'...
[13:45:29] Starting 'task:worker:build'...
[13:45:29] Starting 'task:worker:compile_system'...
[13:45:30] Finished 'task:generator:compile' after 1.1 s
[13:45:30] Starting 'task:generator:copy_deploy'...
src/worker/cache.ts(6,19): error TS2304: Cannot find name 'CacheStorage'.
src/worker/context.ts(9,11): error TS2304: Cannot find name 'CacheStorage'.
src/worker/context.ts(12,17): error TS2304: Cannot find name 'ServiceWorkerRegistration'.
src/worker/manifest-parser.ts(2,20): error TS2307: Cannot find module 'jshashes'.
src/worker/manifest.ts(2,20): error TS2307: Cannot find module 'jshashes'.
src/worker/context.ts(9,11): error TS4031: Public property 'caches' of exported class has or is using private name 'CacheStorage'.
src/worker/context.ts(12,17): error TS4031: Public property 'registration' of exported class has or is using private name 'ServiceWorkerRegistration'.
[13:45:31] TypeScript: 5 semantic errors
[13:45:31] TypeScript: 2 emit errors
[13:45:31] TypeScript: emit failed
[13:45:31] Finished 'task:generator:copy_deploy' after 962 ms
[13:45:31] Finished 'task:generator:build' after 2.06 s
[13:45:31] Finished 'task:worker:compile_system' after 2.06 s
[13:45:31] Starting 'task:worker:bundle'...
Unhandled rejection Error on fetch for dist/src/worker/context.js at file:///Users/crossj/Projects/mobile-toolkit/service-worker/worker/dist/src/worker/context.js
        Loading dist/src/worker/browser_entry.js
        Error: ENOENT: no such file or directory, open '/Users/crossj/Projects/mobile-toolkit/service-worker/worker/dist/src/worker/context.js'
    at Error (native)
```

**Comment by jeffbcross** [comment]
The sanity spec hangs on the push notifications test, waiting for `po.asyncResult` to resolve.

**Comment by jeffbcross** [comment]
The typings issue and sanity spec issue are fixed.

File: service-worker/worker/src/companion/comm.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    // Set a magic value in the message.
+    message['$ngsw'] = true;
+
+    worker.postMessage(message, [channel.port2]);
+    return result;
+  }
+
+  // Send a message to the current controlling worker, waiting for one if needed.
+  private send(message: Object): Observable<any> {
+    let channel = new MessageChannel();
```
Comment (jeffbcross) [inline service-worker/worker/src/companion/comm.ts]: Unused

Comment (alxhub) [inline service-worker/worker/src/companion/comm.ts]: Fixed.

**Comment by alxhub** [comment]
Already merged via another PR.

**Pull request closed without merging**